ouroboros-network-protocols/src/Ouroboros/Network/Protocol/ChainSync/ClientPipelined.hs

```diff
@@ -64,7 +64,8 @@ data ClientPipelinedStIdle n header point tip  m a where
       -> ClientPipelinedStIdle Z header point tip m a
 
     SendMsgRequestNextPipelined
-      :: ClientPipelinedStIdle (S n) header point tip m a
+      :: m ()   -- ^ promptly invoked when 'MsgAwaitReply' is received
+      -> ClientPipelinedStIdle (S n) header point tip m a
```

Comment:
I had noticed this too. I think it can be illuminating to consider the differences.

- `SendMsgRequestNext` is sent _without_ pipelining. Meaning that the sender _can't_ do anything until a reply comes. Thus it's "obvious" for the future sender behavior to explicitly depend on the reply.

- `SenderPipeline` sends the header and then immediately proceeds with subsequent sender behavior (a la pipelining), before the reply even arrives. Thus it's "obvious" for the future sender behavior to explicitly (and crucially) _not_ depend on the reply.

In the second case, the analogous altering of behavior would actually be in the "receiver" data type, but there currently _isn't one_. The ChainSync receiver is "flat", so it's just defined (in terms of the `typed-protocols` constructors) entirely inline within the ChainSync sender.

Reply:
>  - `SendMsgRequestNextPipelined` sends the header and then immediately proceeds with subsequent sender behavior (a la pipelining), before the reply even arrives. Thus it's "obvious" for the future sender behavior to explicitly (and crucially) not depend on the reply.

But you still *could* let the handling of the reply (`recvMsgRoll{For,Back}ward`) depend on whether a `MsgAwaitReply` was received just before even in the pipelined case, right? Eg by providing two `ClientStNext` to `CollectResponse`. (That doesn't subsume the `m ()` in `SendMsgRequestNextPipelined` as that will be invoked promplty once `MsgAwaitReply` arrives).

---

In any case, I don't think this blocking this PR :+1:

Reply:
This is where the `MsgAwaitReply` is observed.

```
chainSyncClientPeerSender n (SendMsgRequestNextPipelined await next) =

    -- pipeline 'MsgRequestNext', the receiver will await for an instruction.
    SenderPipeline
      (ClientAgency TokIdle)
      MsgRequestNext

      (ReceiverAwait (ServerAgency (TokNext TokCanAwait))
        -- await for the reply
        $ \case
          MsgRollForward  header    tip -> ReceiverDone (RollForward header tip)
          MsgRollBackward pRollback tip -> ReceiverDone (RollBackward pRollback tip)

          -- we need to wait for the next message; this time it must come with
          -- an instruction
          MsgAwaitReply -> ReceiverEffect $ do
            await
            pure $ ReceiverAwait (ServerAgency (TokNext TokMustReply)) $ \case
              MsgRollForward  header    tip -> ReceiverDone (RollForward header tip)
              MsgRollBackward pRollback tip -> ReceiverDone (RollBackward pRollback tip))

      (chainSyncClientPeerSender (Succ n) next)
```

> Eg by providing two `ClientStNext` to `CollectResponse`.

And I think what you're suggesting is that the `RollForward` and `RollBackward` ctors would carry a flag (or even a `Maybe x` for some new `x` parameter 😬) indicating whether `MsgAwaitReply` was seen, and then `CollectResponse` would use that flag to choose between the proposed pair of `ClientStNext`s.

I agree that would be more complete information. At this time, though, we have no plans for using it, so we should proceed without it to avoid the noise/confusion.

-----

Good observation 👍

Reply:
> I agree that would be more complete information. At this time, though, we have no plans for using it, so we should proceed without it to avoid the noise/confusion.

Exactly, so "conversely", the only thing that I think *could* be done in this PR would be to *remove* this unnecessary extra flexibility from the synchronous `SendMsgRequestNext`, as mentioned in the last paragraph of https://github.com/IntersectMBO/ouroboros-network/pull/4791#discussion_r1469483545.

Reply:
Have you considered an alternative approach which would make things more symmetric with respect to the pipelined - non-pipelined messages, i.e. extend the `ChainSyncInstruction` type:

```hs
data ChainSyncInstruction header point tip
    = RollForward  !header !tip
    | RollBackward !point  !tip
    | AwaitInstruction
```

and add a callback to `ClientStNext` (the drawback is that it's also used in `
```hs
data ClientStNext n header point tip m a =
     ClientStNext {
       -- | Callback for 'MsgRollForward' message.
       --
       recvMsgRollForward
         :: header
         -> tip
         -> m (ClientPipelinedStIdle n header point tip m a),

       -- | Callback for 'MsgRollBackward' message.
       --
       recvMsgRollBackward
         :: point
         -> tip
         -> m (ClientPipelinedStIdle n header point tip m a),
         
       -- | callback for 'MsgAwaitReply' message.
       --
       recvMsgAwaitResponse
         :: m (ClientStNext n header point tip m a)
     }
```

* `recvMsgAwaitRespoinse` would be called when the pipelining client collects the message, rather than when it is received - would this be acceptable?
* `recvMsgAwaitRespoinse` is not always needed, this could be fixed by using a different type for blocking / non-blocking requests

Reply:
> would be called when the pipelining client collects the message, rather than when it is received - would this be acceptable?

No; the straight forward design relies on prompt notification. As soon as the `MsgAwaitReply` arrives, we know that the peer has no more headers to offer us. We wouldn't want to wait until they do have one more, eg.

Reply:
> I agree that would be more complete information. At this time, though, we have no plans for using it, so we should proceed without it to avoid the noise/confusion.

We may have a use case for it, in the LoP implementation. Namely, we may want to pause the leaking bucket when `MsgAwaitReply` is received, then resume the leaking when the next message is sent. The straightforward way to resume the leaking would be to pass two callbacks to `SendMsgRequestNext*`.

Reply:
Thanks @nbacquey . There's a subtlety, though.

The thing my comment motivates removing is only available for `SendMsgRequestNext`, _not_ for `SendMsgRequestNextPipeline`.

- `(ClientNext ..., m (ClientNext ...))` works for `SendMsgRequestNext`.
- `SendMsgRequestNextPipelined` _must_ receive the `m ()` part, since it's the one that processes `MsgAwaitReply` and we want that to be handled promptly.
- `Collect...` is currently what specifies the `ClientNext` part. It should _not_ have a `m ClientNext`, since it was not necessarily invoked before the `MsgAwaitReply` arrived.

We could change it so that the mini protocol logic must provide the handler for the messages at the same time that it requests them. IE move the `ClientNext` logic from `Collect...` to `SendMsgRequestNextPipeline`. But then we'd have to hold that in the queue, as the responses. Might ultimately be simpler 🤷. I tinkered with that very briefly, but the types got a bit confusing: the `n` in the type of `ClientNext` is currently the number of `Send..Pipeline`s that don't have matching `Collect`s---so the type of the elements in the queue would vary with their depth in the queue? Seems wrong. ... perhaps the proposed `ClientNext` argument to `Send...Pipeline` would need to be polymorphic in `n`?

My point is: this seems like a significantly deeper redesign than the current state of this PR. And as Esgen originally explained the current PR is enough: the `m ()` can update a `TVar` and so on.